### PR TITLE
ARROW-3070: [Packaging] Use Bintray

### PR DESCRIPTION
**Pull request by kou**
dev/release/03-binary.sh uploads binaries (.deb, .rpm, .whl and conda
.tar.bz2) to https://bintray.com/apache/arrow/ . It uses debian-rc,
ubuntu-rc, centos-rc and python-rc packages on Bintray.

dev/release/03-binary.sh signs and generates checksums binaries. This
ignores signs and checksums generated by crossbow. Because we need
some special signing. For example, .dsc and .changes should use inline
sign instead of separated sign file. debsign command does this. .rpm
should be signed by "rpm --addsign".

dev/release/03-binary.sh also generates metadata to install .deb and
.rpm files by apt and yum. See the changes for site/install.md how to
use https://bintray.com/apache/arrow/ as APT and Yum repository.

This means that this change includes work for
https://issues.apache.org/jira/browse/ARROW-1349 .

We can install not only RC binaries but also released binaries by apt
and yum. It will help us to verify binaries.

dev/release/post-02-binary.sh downloads verified binaries at
debian-rc, ubuntu-rc, centos-rc and python-rc packages on Bintray and
uploads them to debian, ubuntu, centos and python packages on Bintray.

APT and Yum repositories for RC and release are separated. So RC
doesn't affect users.

Release manager needs write access to https://bintray.com/apache/arrow .

### CODE REVIEW

**Comment by kou** [comment]
I confirmed that this almost works well. But my Bintray account exceeded API limit. So I need to finish more tests. This is the reason why this pull request marks as WIP.

I've uploaded binaries for 0.11.0-rc1 and 0.11.0-rc0 to https://bintray.com/apache/arrow . We can confirm the result for `dev/release/03-binary.sh`.

**Comment by kou** [comment]
`03-binary.sh` needs to use some special commands such as `debsign`, `apt-ftparchive`, `rpm` and `createrepo`. So I also add `Dockerfile` to run them. It'll work on macOS but I didn't try yet.

`03-binary.sh` requires recent GnuPG (maybe GnuPG >= 2.1). Because it uses `gpg-agent` to sign in Docker. It will require recent GnuPG.

**Comment by kou** [comment]
@kszucs I think that we move signing and checksum generation features to `dev/release/03-binary.sh` from crossbow because this task needs to run many external commands. So shell script may be better.

If you think this should be implemented in crossbow, we can work on it as follow-up pull request.

**Comment by kszucs** [comment]
Thanks @kou for doing this! It looks good, let me know when and can I test it :)

> In general I prefer python (mainly because of portability and testability, at some point We'll need to mock these APIs), but of course We can do that as a follow-up.

**Comment by kou** [comment]
Now, this pull request is ready to review.

I've uploaded binaries for 0.11.0-rc1, 0.11.0, 0.11.1-rc0 and 0.11.1:

For Debian GNU/Linux:

  * 0.11.0-rc1: https://bintray.com/apache/arrow/debian-rc/0.11.0-rc1#files
  * 0.11.0: https://bintray.com/apache/arrow/debian/0.11.01#files
  * 0.11.1-rc0: https://bintray.com/apache/arrow/debian-rc/0.11.1-rc0#files
  * 0.11.1: https://bintray.com/apache/arrow/debian/0.11.1#files

For Ubuntu:

  * 0.11.0-rc1: https://bintray.com/apache/arrow/ubuntu-rc/0.11.0-rc1#files
  * 0.11.0: https://bintray.com/apache/arrow/ubuntu/0.11.01#files
  * 0.11.1-rc0: https://bintray.com/apache/arrow/ubuntu-rc/0.11.1-rc0#files
  * 0.11.1: https://bintray.com/apache/arrow/ubuntu/0.11.1#files

For CentOS:

  * 0.11.0-rc1: https://bintray.com/apache/arrow/centos-rc/0.11.0-rc1#files
  * 0.11.0: https://bintray.com/apache/arrow/centos/0.11.01#files
  * 0.11.1-rc0: https://bintray.com/apache/arrow/centos-rc/0.11.1-rc0#files
  * 0.11.1: https://bintray.com/apache/arrow/centos/0.11.1#files

For Python:

  * 0.11.0-rc1: https://bintray.com/apache/arrow/python-rc/0.11.0-rc1#files
  * 0.11.0: https://bintray.com/apache/arrow/python/0.11.01#files
  * 0.11.1-rc0: https://bintray.com/apache/arrow/python-rc/0.11.1-rc0#files
  * 0.11.1: https://bintray.com/apache/arrow/python/0.11.1#files

We can test the uploaded binaries for Debian GNU/Linux and Ubuntu by `apt`.

For RC (I'll write this to Wiki when this pull request is merged):

```shell
sudo apt update
sudo apt install -y -V apt-transport-https lsb-release
sudo wget -O /usr/share/keyrings/apache-arrow-keyring.gpg https://dl.bintray.com/apache/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')-rc/apache-arrow-keyring.gpg
sudo tee /etc/apt/sources.list.d/apache-arrow.list <<APT_LINE
deb [arch=amd64 signed-by=/usr/share/keyrings/apache-arrow-keyring.gpg] https://dl.bintray.com/apache/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')-rc/ $(lsb_release --codename --short) main
deb-src [signed-by=/usr/share/keyrings/apache-arrow-keyring.gpg] https://dl.bintray.com/apache/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')-rc/ $(lsb_release --codename --short) main
APT_LINE
sudo apt update
sudo apt install -y -V libarrow-dev # For C++
sudo apt install -y -V libarrow-glib-dev # For GLib (C)
sudo apt install -y -V libparquet-dev # For Apache Parquet C++
sudo apt install -y -V libparquet-glib-dev # For Parquet GLib (C)
```

For release (This is written in `site/install.md`):

```shell
sudo apt update
sudo apt install -y -V apt-transport-https lsb-release
sudo wget -O /usr/share/keyrings/apache-arrow-keyring.gpg https://dl.bintray.com/apache/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')/apache-arrow-keyring.gpg
sudo tee /etc/apt/sources.list.d/apache-arrow.list <<APT_LINE
deb [arch=amd64 signed-by=/usr/share/keyrings/apache-arrow-keyring.gpg] https://dl.bintray.com/apache/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')/ $(lsb_release --codename --short) main
deb-src [signed-by=/usr/share/keyrings/apache-arrow-keyring.gpg] https://dl.bintray.com/apache/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')/ $(lsb_release --codename --short) main
APT_LINE
sudo apt update
sudo apt install -y -V libarrow-dev # For C++
sudo apt install -y -V libarrow-glib-dev # For GLib (C)
sudo apt install -y -V libparquet-dev # For Apache Parquet C++
sudo apt install -y -V libparquet-glib-dev # For Parquet GLib (C)
```

We can test the uploaded binaries for CentOS by `yum`:

For RC (I'll write this to Wiki when this pull request is merged):

```shell
sudo tee /etc/yum.repos.d/Apache-Arrow.repo <<REPO
[apache-arrow]
name=Apache Arrow
baseurl=https://dl.bintray.com/apache/arrow/centos-rc/\$releasever/\$basearch/
gpgcheck=1
enabled=1
gpgkey=https://dl.bintray.com/apache/arrow/centos-rc/RPM-GPG-KEY-apache-arrow
REPO
sudo yum install -y epel-release
sudo yum install -y --enablerepo=epel arrow-devel # For C++
sudo yum install -y --enablerepo=epel arrow-glib-devel # For GLib (C)
sudo yum install -y --enablerepo=epel parquet-devel # For Apache Parquet C++
sudo yum install -y --enablerepo=epel parquet-glib-devel # For Parquet GLib (C)
```

For release (This is written in `site/install.md`):

```shell
sudo tee /etc/yum.repos.d/Apache-Arrow.repo <<REPO
[apache-arrow]
name=Apache Arrow
baseurl=https://dl.bintray.com/apache/arrow/centos/\$releasever/\$basearch/
gpgcheck=1
enabled=1
gpgkey=https://dl.bintray.com/apache/arrow/centos/RPM-GPG-KEY-apache-arrow
REPO
sudo yum install -y epel-release
sudo yum install -y --enablerepo=epel arrow-devel # For C++
sudo yum install -y --enablerepo=epel arrow-glib-devel # For GLib (C)
sudo yum install -y --enablerepo=epel parquet-devel # For Apache Parquet C++
sudo yum install -y --enablerepo=epel parquet-glib-devel # For Parquet GLib (C)
```

We can upload binaries for RC by the following command line (I'll write this to Wiki when this pull is merged):

```console
% BINTRAY_USER=${YOUR_BINTRAY_USER} BINTRAY_PASSWORD=${YOUR_BINTRAY_API_KEY} dev/release/03-binary.sh ${VERSION} ${RC} ${YOUR_GPG_KEY_ID} ${CROSSBOW_SIGN_OUTPUT_DIR}
```

Example for uploading binaries for 0.11.0 RC1 by me:

```console
% BINTRAY_USER=kou BINTRAY_PASSWORD=secret dev/release/03-binary.sh 0.11.0 1 08D3564B7C6A9CAFBFF6A66791D18FCF079F8007 packages/build-38
```

We can upload binaries for voted release by the following command line (I'll write this to Wiki when this pull is merged):

```console
% BINTRAY_USER=${YOUR_BINTRAY_USER} BINTRAY_PASSWORD=${YOUR_BINTRAY_API_KEY} dev/release/post-02-binary.sh ${VERSION} ${RC}
```

Example for uploading binaries for 0.11.0 by me:

```console
% BINTRAY_USER=kou BINTRAY_PASSWORD=secret dev/release/post-02-binary.sh 0.11.0 1
```

You can try `03-binary.sh` and `post-02-binary.sh` by removing uploaded binaries. Because we don't provide binaries from bintray.com.

Here is a command to remove uploaded binaries of 0.11.0 RC1 for Debian GNU/Linux:

```console
% curl --verbose --fail --basic --user ${YOUR_BINTRAY_USER}:${YOUR_BINTRAY_API_KEY} --header 'Content-Type: application/json' --request DELETE https://bintray.com/api/v1/packages/apache/arrow/debian-rc/versions/0.11.0
```

Here is a command to remove uploaded binaries of 0.11.1 for Python:

```console
% curl --verbose --fail --basic --user ${YOUR_BINTRAY_USER}:${YOUR_BINTRAY_API_KEY} --header 'Content-Type: application/json' --request DELETE https://bintray.com/api/v1/packages/apache/arrow/python/versions/0.11.1
```

Note that you can't try these scripts many times in a day because you can only do 1440 API calls per day. both `03-binary.sh` and `post-02-binary.sh` use about 500 API calls.

**Comment by kszucs** [comment]
@kou it looks pretty cool, however the bintray repo is hardcoded to apache/arrow. Perhaps could You add another env variable to let me try it out with a non-apache bintray repo?

**Comment by kou** [comment]
It makes sense.
I've added `BINTRAY_REPOSITORY` environment variable.

Now, you can try this with your Bintray repository.

Here is an example to use https://bintray.com/kou/arrow :

```console
% BINTRAY_REPOISITORY=kou/arrow ...
```

Note that you need to create the following "package"s in your repository:

  * debian
  * debian-rc
  * ubuntu
  * ubuntu-rc
  * centos
  * centos-rc
  * python
  * python-rc

See also: https://bintray.com/kou/arrow

**Comment by kszucs** [comment]
Thanks @kou! I'll try this out today.

File: dev/release/03-binary.sh [hunk 1]
```diff
@@ -50,26 +50,467 @@ if [ ! -d "$artifact_dir" ]; then
...
+    --user ${uid}:${gid} \
+    --volume "$PWD":/host \
+    ${docker_image_name} \
+    bash -c "cd /host && $*"
+}
+
+docker_run_gpg_ready() {
+  local gpg_agent_socket_dir="$(gpgconf --list-dir socketdir)"
+  local uid=$(id -u)
+  local gid=$(id -g)
```
Comment (kszucs) [inline dev/release/03-binary.sh]: On osx it's 20 (staff:*:20:root) for me, which already exists in the docker container so groupadd fails with `groupadd: GID '20' already exists`. If I set `gid=uid` it works.

**Comment by kszucs** [comment]
@kou It's still uploading the binaries, but it works great (with the two modifications I've mentioned above). I still need to try the `post-02-binary.sh` script, just waiting for the network...

**Comment by kszucs** [comment]
Post release script works too: https://bintray.com/kszucs/arrow
If We can address the OSX problems above this PR is ready to go!

File: dev/release/03-binary.sh [hunk 2]
```diff
@@ -50,26 +50,467 @@ if [ ! -d "$artifact_dir" ]; then
...
+  commands="${commands} && chown ${USER}: /run/user/${uid}"
+  commands="${commands} && cd /host"
+  commands="${commands} && sudo -u ${USER} -H $*"
+  docker \
+    run \
+    --rm \
+    --tty \
+    --interactive \
+    --volume "$PWD":/host \
+    --volume "${HOME}/.gnupg:/home/${USER}/.gnupg:ro" \
```
[the remark this answers is not in the document]
Comment (kou) [reply dev/release/03-binary.sh]: We should not run `gpg-agent` in Docker. We should use `gpg-agent` running on host in Docker.

**Comment by kou** [comment]
I changed `gpg-agent` related logic to use `gpg-agent` over SSH instead of sharing `gpg-agent` sockets by Docker volume.
Now, it works on both GNU/Linux and macOS.

**Comment by wesm** [comment]
@kszucs ?

**Comment by kszucs** [comment]
It indeed works nicely on OSX too! Created an [issue](https://issues.apache.org/jira/browse/ARROW-3867) for the documentation.

Thanks @kou!

**Review by kszucs — APPROVED** [review]

**Pull request closed without merging**